### PR TITLE
[WIP, SFT] Support context parallelism for SFT

### PR DESCRIPTION
**Pull request by xingyaoww**

### CODE REVIEW

**Comment by xingyaoww** [comment]
ok the training script is working now!

image

**Comment by xingyaoww** [comment]
This PR should be ready for review! I just added a CI that checks for loss match

image

File: verl/trainer/fsdp_sft_trainer.py [hunk 1]
```diff
@@ -165,6 +193,14 @@ def _build_model_optimizer(self):
         trust_remote_code = self.config.model.trust_remote_code
         # load config first
         config = AutoConfig.from_pretrained(local_model_path, trust_remote_code=trust_remote_code)
+        if self.use_remove_padding:
+            assert self.config.ulysses_sequence_parallel_size > 1, "Remove padding is only supported with sequence parallel"
```
Comment (vermouth1992) [inline verl/trainer/fsdp_sft_trainer.py]: I assume it should be the opposite? Sequence parallel is only support when remove_padding is enabled?

File: verl/trainer/fsdp_sft_trainer.py [hunk 2]
```diff
@@ -289,8 +404,12 @@ def training_step(self, batch: TensorDict):
         n_micro_batches = len(micro_batches)
         step_loss = 0
         for micro_batch in micro_batches:
-            loss = self._compute_loss(batch=micro_batch) / n_micro_batches
-            loss.backward()
+            if self.use_remove_padding and self.config.ulysses_sequence_parallel_size > 1:
+                # micro_batch = micro_batch.to('cuda')
+                loss = self._compute_loss_and_backward_sp(batch=micro_batch) / n_micro_batches
```
Comment (vermouth1992) [inline verl/trainer/fsdp_sft_trainer.py]: Can we combine the two functions as there are plenty replicated code?

**Comment by vermouth1992** [comment]
Nice work!

**Comment by vermouth1992** [comment]
Could you run the formatting script?

**Comment by xingyaoww** [comment]
done! @vermouth1992

**Comment by vermouth1992** [comment]
There are conflicts with newly merge MR..

**Review by vermouth1992 — APPROVED** [review]
